### DISCUSSION TITLE
Duda backtracking

### DISCUSSION THREAD

**Discussion by IgnacioHqz** [IIC2133-PUC/2024-2#168] [Preguntas Materia] [answered]
Hola, estudiando la materia de backtracking me surgió una udda con respecto al conjunto de variables que se utilizan en el ejemplo del laberinto.
Cómo sería el conjunto de variables si a priori no se sabe cuantas variables se van a utilizar debido a que no se sabe cuántas posiciones tendrá el camino que conduce a la salida? Me genera dudas eso, ya que si no sabemos cuantas variables usaremos, cómo se podría definir un conjunto de vairables sin asignar? Si me pudieran aclarar esto lo agradecería mucho ya que me confundió un poco.
Muchas gracias!

![consulta](https://github.com/user-attachments/assets/7bd2cbf0-a084-4b3a-83db-4664b5eac225)

**Comment by Aleksandsk** [comment, marked as answer]
El conjunto de variables X lo puedes interpretar como el conjunto de todas las posiciones con su respectiva etiqueta nonvisited, visited o exit (puedes pensarlo como una matriz donde en cada X[i][j] tienes la etiqueta de la posición i,j). 

El pseudocódigo conserva en X todo el estado del laberinto producto de las asignaciones anteriores, es decir, sabes las variables que has visitado o no. 

En cada llamado recursivo, isSolvable recibe como argumento X y la nueva variable x a revisar, y utiliza X para verificar si la variable x recibida fue visitada o no, para no revisarla de nuevo. Solo retornara true cuando haya encontrado la variable etiquetada como exit, y esto ocurre solo cuando pasas por posiciones nonvisited hasta llegar a exit. 

Cuando llegas a una posición visited, te devuelves para seguir el camino por una posición nonvisited, lo que significa volver a etiquetarla como nonvisited, es decir, reviertes la asignación anterior (Linea 9).

Entonces no necesitas saber el largo del camino que conduce a la salida porque no influye en la búsqueda y tampoco necesitas explícitamente un conjunto de variables sin asignar porque necesitas saber las variables que asignaste antes y en caso de equivocarte, reviertes las asignaciones, y buscas por otro lado.

Espero que haya quedado mas claro y cualquier cosa pregunta nomas!

**Reply by IgnacioHqz** [reply]
Perfecto me quedó más claro ahora, muchas gracias!
